File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/resolution/ResolutionResult.java
package com.foundgine.core.semantic.resolution;

import java.util.List;

public final class ResolutionResult {
	private final ResolutionOutcome outcome;
	private final ResolvedReference resolved;
	private final String unresolvedReason;
	private final List<ResolutionEvidence> evidence;

	private ResolutionResult(ResolutionOutcome outcome, ResolvedReference resolved, String reason,
			List<ResolutionEvidence> evidence) {
		this.outcome = outcome;
		this.resolved = resolved;
		this.unresolvedReason = reason;
		this.evidence = evidence == null ? List.of() : List.copyOf(evidence);
	}

	public static ResolutionResult success(ResolvedReference reference) {
		return new ResolutionResult(ResolutionOutcome.RESOLVED, reference, null, reference.evidence());
	}

	public static ResolutionResult ambiguous(String reason, List<ResolutionEvidence> evidence) {
		return new ResolutionResult(ResolutionOutcome.AMBIGUOUS, null, reason, evidence);
	}

	public static ResolutionResult notFound(String reason, List<ResolutionEvidence> evidence) {
		return new ResolutionResult(ResolutionOutcome.NOT_FOUND, null, reason, evidence);
	}

	public ResolutionOutcome outcome() {
		return outcome;
	}

	public ResolvedReference resolved() {
		return resolved;
	}

	public String unresolvedReason() {
		return unresolvedReason;
	}

	public List<ResolutionEvidence> evidence() {
		return evidence;
	}
}

File: src/csharp/Foundgine.Core/Semantic/Resolution/ResolutionResult.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Resolution;

public enum CandidateEvidenceKind : byte
{
    ExactIdentity,
    Alias,
    Trigram,
    FullText,
    Bm25,
    GraphSimilarity,
    VectorSimilarity,
    Relationship
}

public sealed record ResolutionEvidence(
    string Description,
    CandidateEvidenceKind? Kind = null,
    double? Score = null);

public enum ResolutionOutcome
{
    Resolved,
    Ambiguous,
    NotFound
}

public sealed record ResolvedReference(
    EntityId EntityType,
    string IdentityValue,
    double Confidence,
    string Reason,
    IReadOnlyList<ResolutionEvidence> Evidence);

public sealed record ResolutionResult
{
    public ResolutionOutcome Outcome { get; }
    public ResolvedReference? Resolved { get; }
    public string? UnresolvedReason { get; }
    public IReadOnlyList<ResolutionEvidence> Evidence { get; }

    private ResolutionResult(
        ResolutionOutcome outcome,
        ResolvedReference? resolved,
        string? reason,
        IReadOnlyList<ResolutionEvidence> evidence)
    {
        Outcome = outcome;
        Resolved = resolved;
        UnresolvedReason = reason;
        Evidence = evidence;
    }

    public static ResolutionResult Success(ResolvedReference reference) =>
        new(ResolutionOutcome.Resolved, reference, null, reference.Evidence);

    public static ResolutionResult Ambiguous(
        string reason,
        IReadOnlyList<ResolutionEvidence> evidence) =>
        new(ResolutionOutcome.Ambiguous, null, reason, evidence);

    public static ResolutionResult NotFound(
        string reason,
        IReadOnlyList<ResolutionEvidence> evidence) =>
        new(ResolutionOutcome.NotFound, null, reason, evidence);
}
